src/protocol/sort/secureapplyinv.rs

```diff
@@ -0,0 +1,171 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::fabric::Network,
+    protocol::{
+        context::ProtocolContext,
+        reveal::reveal,
+        sort::ApplyInvStep::{RevealPermutation, ShuffleInputs, ShufflePermutation},
+        RecordId,
+    },
+    secret_sharing::Replicated,
+};
+use embed_doc_image::embed_doc_image;
+use futures::future::try_join_all;
+use permutation::Permutation;
+
+use super::{
+    apply::apply,
+    shuffle::{generate_random_permutation, Shuffle},
+};
+use tokio::try_join;
+
+/// This is APPLYINV(Algorithm 4) described in <https://eprint.iacr.org/2019/695.pdf>.
+/// This is a protocol that applies the inverse of a shared permutation to a shared-vector representation.
+/// Input: Each helpers know their own secret shares of input and permutation
+/// Output: At the end of the protocol, all helpers receive inputs after the permutation is applied
+#[derive(Debug)]
+#[embed_doc_image("secureapplyinv", "images/sort/secureapplyinv.png")]
+pub struct SecureApplyInv<'a, F> {
+    input: &'a mut Vec<Replicated<F>>,
+    permutation: &'a mut Vec<Replicated<F>>,
+}
+
+impl<'a, F: Field> SecureApplyInv<'a, F> {
+    #[allow(dead_code)]
+    pub fn new(input: &'a mut Vec<Replicated<F>>, permutation: &'a mut Vec<Replicated<F>>) -> Self {
+        Self { input, permutation }
+    }
+    /// This algorithm applies a permutation on the inputs. The permutation is not available in clear and
+    /// we do not want to reveal the actual permutation to any of the helpers while applying it.
+    /// Steps
+    /// ![Secure Apply Inv steps][secureapplyinv]
+    /// 1. Generate a permutation
+    /// 2. Secret shared permutation is shuffled with a permutation
+    /// 3. Secret shared value is shuffled using the same permutation as 1
+    /// 4. The permutation is revealed
+    /// 5. All helpers call `apply` to apply the permutation locally.
+    #[allow(dead_code)]
+    pub async fn execute<N: Network>(
+        &mut self,
+        ctx: &ProtocolContext<'_, N>,
+    ) -> Result<(), BoxError> {
+        // 1. Generate permutation
+        let permutations = generate_random_permutation(self.input.len(), &ctx.prss());
+        // 2. Shuffle permutations
+        let mut shuffle_permutation = Shuffle::new(self.permutation);
+        let h0_future = shuffle_permutation.execute(ctx.narrow(&ShufflePermutation), &permutations);
+
+        // 3. Shuffle inputs
+        let mut shuffle_input = Shuffle::new(self.input);
+        let h1_future = shuffle_input.execute(ctx.narrow(&ShuffleInputs), &permutations);
+
+        try_join!(h0_future, h1_future).unwrap();
```

Comment:
I think it'll be more readable to compress this into just 4 lines of code with a lot less local variable names.
It also starts to show just how confusing it is to have both a local variable `permutations` which is actually very different from `self.permutation`. So I suggest a slightly different naming scheme.

```
let (shuffled_inputs, shuffled_permutation) = try_join(
    Shuffle::new(self.input).execute(ctx.narrow(&ShuffleInputs), &random_permutations),
    Shuffle::new(self.sort_permutation).execute(ctx.narrow(&ShufflePermutation), &random_permutation),
).await?;

Reply:
I had tried to do that but rust wasnt happy to create a variable and call a function in the same line :( temporary value dropped while borrowed
consider using a `let` binding to create a longer lived value